Loading capture * .gfxtrace, report: panic: runtime error: index out of range

```
I20170324-093527524[main][models.Capture.loadCapture] Loading capture /ws/gpu/training_20170323_1156.gfxtrace...
W20170324-093528956[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: panic: runtime error: index out of range
W20170324-093528956[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis:
W20170324-093528957[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: goroutine 75 [running]:
W20170324-093528957[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/gfxapi/gles.moveClientVBsToVAs(0x766d360, 0xc422357e90, 0xc421aec190, 0xc422ba9d70, 0x0, 0xffffffffffffffff, 0x297, 0x7672c40, 0xc4205326f0, 0xc4200ae310, ...)
W20170324-093528957[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/gfxapi/gles/compat.go:1071 +0x611
W20170324-093528957[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/gfxapi/gles.compat.func4(0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9f20)
W20170324-093528958[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/gfxapi/gles/compat.go:521 +0x15fd
W20170324-093528958[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.transform.Transform(0x5fbd6d9, 0x6, 0xc420659f40, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9f20)
W20170324-093528958[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/atom/transform/transforms.go:75 +0x75
W20170324-093528958[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.(*transform).Transform(0xc421c9dc00, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9f20)
W20170324-093528958[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	<autogenerated>:10 +0xb1
W20170324-093528961[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.TransformWriter.MutateAndWrite(0xc4200ae310, 0x7613120, 0xc421c9dc00, 0x76130e0, 0xc422ba9f20, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0)
W20170324-093528961[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/atom/transform/transforms.go:98 +0x83
W20170324-093528961[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.(*TransformWriter).MutateAndWrite(0xc422ba9f50, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0)
W20170324-093528961[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	<autogenerated>:8 +0xa1
W20170324-093528961[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/gfxapi/gles.(*readFramebuffer).Transform(0xc422bae090, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9f50)
W20170324-093528961[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/gfxapi/gles/read_framebuffer.go:45 +0x73
W20170324-093528962[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.TransformWriter.MutateAndWrite(0xc4200ae310, 0x76111e0, 0xc422bae090, 0x76130e0, 0xc422ba9f50, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0)
W20170324-093528962[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/atom/transform/transforms.go:98 +0x83
W20170324-093528962[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.(*TransformWriter).MutateAndWrite(0xc422ba9f80, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0)
W20170324-093528962[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	<autogenerated>:8 +0xa1
W20170324-093528963[Thread-1][server.GapisConnection$GRpcGapisConnection$LogMonitorThread.run] getLogStream() threw unexpected exception
W20170324-093528963[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/gfxapi/gles.undefinedFramebuffer.func1(0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9f80)
W20170324-093528963[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/gfxapi/gles/undefined_framebuffer.go:32 +0x9c
W20170324-093528963[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.transform.Transform(0x5fca805, 0x10, 0xc421c9da20, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9f80)
W20170324-093528964[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/atom/transform/transforms.go:75 +0x75
W20170324-093528964[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.(*transform).Transform(0xc421c9da40, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9f80)
W20170324-093528964[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	<autogenerated>:10 +0xb1
W20170324-093528965[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.TransformWriter.MutateAndWrite(0xc4200ae310, 0x7613120, 0xc421c9da40, 0x76130e0, 0xc422ba9f80, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] LoadData error
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: UNAVAILABLE: End of stream or IOException
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:500)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:459)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture$TrustedFuture.get(AbstractFuture.java:76)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:142)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.gapid.rpclib.rpccore.Rpc$Result.get(Rpc.java:148)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.gapid.models.CaptureDependentModel.processResult(CaptureDependentModel.java:106)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.gapid.models.CaptureDependentModel$2.onRpcThread(CaptureDependentModel.java:84)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.gapid.models.CaptureDependentModel$2.onRpcThread(CaptureDependentModel.java:81)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.gapid.util.UiCallback.onFinish(UiCallback.java:46)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.gapid.rpclib.rpccore.Rpc$1.run(Rpc.java:100)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:435)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:900)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:811)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:675)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:112)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:435)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:900)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:811)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:675)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:112)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:435)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:900)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:811)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:675)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:466)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:442)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:481)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:398)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:513)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:52)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.internal.SerializingExecutor$TaskRunner.run(SerializingExecutor.java:154)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at java.lang.Thread.run(Thread.java:745)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: End of stream or IOException
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	at io.grpc.Status.asRuntimeException(Status.java:545)
E20170324-093528963[grpc-default-executor-2][models.CaptureDependentModel.processResult] 	... 9 more
W20170324-093528963[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/atom/transform/transforms.go:98 +0x83
W20170324-093529021[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.(*TransformWriter).MutateAndWrite(0xc422ba9fb0, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0)
W20170324-093529022[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	<autogenerated>:8 +0xa1
W20170324-093529022[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/gfxapi/gles.(*findIssues).Transform(0xc422ba9bc0, 0x766d360, 0xc422357e90, 0x297, 0x7672c40, 0xc4205326f0, 0x76130e0, 0xc422ba9fb0)
W20170324-093529022[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/gfxapi/gles/find_issues.go:102 +0x168
W20170324-093529022[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.TransformWriter.MutateAndWrite(0xc4200ae310, 0x76111a0, 0xc422ba9bc0, 0x76130e0, 0xc422ba9fb0, 0x766d360, 0xc4220ff380, 0x297, 0x7672c40, 0xc4205326f0)
W20170324-093529040[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/atom/transform/transforms.go:98 +0x83
W20170324-093529041[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.(*TransformWriter).MutateAndWrite(0xc422b7e000, 0x766d360, 0xc4220ff380, 0x297, 0x7672c40, 0xc4205326f0)
W20170324-093529041[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	<autogenerated>:8 +0xa1
W20170324-093529041[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/atom/transform.Transforms.Transform(0xc42053bd80, 0x5, 0x8, 0x766d360, 0xc4220ff380, 0xc420caa000, 0x3987, 0x3987, 0x76113a0, 0xc421eb64c0)
W20170324-093529041[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/atom/transform/transforms.go:40 +0x1ba
W20170324-093529041[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/gfxapi/gles.api.Replay(0x766d360, 0xc4220fed20, 0xc421eba008, 0xc421eba010, 0x5c2d8a0, 0x776cbe0, 0xc422548100, 0x1, 0x1, 0x76cfda0, ...)
W20170324-093529042[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/gfxapi/gles/replay.go:211 +0xba3
W20170324-093529042[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/gfxapi/gles.(*api).Replay(0x776cbe0, 0x766d360, 0xc4220fed20, 0xc421eba008, 0xc421eba010, 0x5c2d8a0, 0x776cbe0, 0xc422548100, 0x1, 0x1, ...)
W20170324-093529042[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	<autogenerated>:2420 +0xfb
W20170324-093529042[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/replay.(*batcher).send(0xc42214a930, 0x766d360, 0xc4220fed20, 0xc422548100, 0x1, 0x1, 0x0, 0x0)
W20170324-093529042[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/replay/batcher.go:178 +0x934
W20170324-093529042[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: github.com/google/gapid/gapis/replay.(*batcher).run(0xc42214a930, 0x766d360, 0xc422160d80)
W20170324-093529042[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/replay/batcher.go:95 +0x5f9
W20170324-093529043[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: created by github.com/google/gapid/gapis/replay.(*Manager).getBatchStream
W20170324-093529043[server.ChildProcess$LoggingStringHandler][server.ChildProcess$LoggingStringHandler.lambda$new$71] gapis: 	/myenv/go/src/github.com/google/gapid/gapis/replay/manager.go:55 +0x3f9
W20170324-093529043[server.ChildProcess-gapis][server.ChildProcess.onExit] The gapis process exited with a non-zero exit value: 2
I20170324-093529043[server.ChildProcess-gapis][server.ChildProcess.shutdown] Shutting down gapis
```

--- a/gapii/cc/gles_spy_externs.cpp
+++ b/gapii/cc/gles_spy_externs.cpp
@@ -29,6 +29,9 @@ MsgID GlesSpy::newMsg(CallObserver*, uint32_t, const char*) { return 0; }
 void GlesSpy::addTag(CallObserver*, uint32_t, const char*) {}
 
 u32Limits GlesSpy::IndexLimits(CallObserver*, uint8_t* indices, uint32_t indices_type, uint32_t offset, uint32_t count) {
+    if (count == 0) {
+        return u32Limits(0, 0);
+    }
     uint32_t low = ~(uint32_t)0;
     uint32_t high = 0;
     switch (indices_type) {
@@ -57,7 +60,7 @@ u32Limits GlesSpy::IndexLimits(CallObserver*, uint8_t* indices, uint32_t indices
             break;
         }
     }
-    return u32Limits(low, high);
+    return u32Limits(low, high+1-low);
 }
 
 void GlesSpy::onGlError(CallObserver* observer, GLenum_Error err) {

--- a/gapis/gfxapi/gles/api/draw_commands.api
+++ b/gapis/gfxapi/gles/api/draw_commands.api
@@ -110,7 +110,7 @@ sub void DrawElements(ref!Context    ctx,
 
       // Read the vertices
       limits := IndexLimits(as!u8*(index_data), indices_type, offset, count)
-      ReadVertexArrays(ctx, limits.first + as!u32(base_vertex), (limits.last - limits.first) + 1, as!u32(instance_count))
+      ReadVertexArrays(ctx, limits.first + as!u32(base_vertex), limits.count, as!u32(instance_count))
 
       // No need to read the indices - they're already in a GPU-side buffer.
     } else {
@@ -119,7 +119,7 @@ sub void DrawElements(ref!Context    ctx,
 
       // Read the vertices
       limits := IndexLimits(index_data, indices_type, 0, count)
-      ReadVertexArrays(ctx, limits.first + as!u32(base_vertex), (limits.last - limits.first) + 1, as!u32(instance_count))
+      ReadVertexArrays(ctx, limits.first + as!u32(base_vertex), limits.count, as!u32(instance_count))
 
       // Read the indices
       read(index_data[0:count * IndexSize(indices_type)])

--- a/gapis/gfxapi/gles/compat.go
+++ b/gapis/gfxapi/gles/compat.go
@@ -514,11 +514,11 @@ func compat(ctx context.Context, device *device.Instance) (transform.Transformer
 		case *GlDrawArrays:
 			if target.vertexArrayObjects == required {
 				if clientVAsBound(c, clientVAs) {
-					first := int(a.FirstIndex)
-					last := first + int(a.IndicesCount) - 1
+					first := uint32(a.FirstIndex)
+					count := uint32(a.IndicesCount)
 					t := newTweaker(ctx, out)
 					defer t.revert()
-					moveClientVBsToVAs(ctx, t, clientVAs, first, last, i, a, s, c, out)
+					moveClientVBsToVAs(ctx, t, clientVAs, first, count, i, a, s, c, out)
 				}
 			}
 
@@ -553,7 +553,7 @@ func compat(ctx context.Context, device *device.Instance) (transform.Transformer
 						// application pool.
 						a.Extras().Observations().ApplyReads(s.Memory[memory.ApplicationPool])
 						limits := e.calcIndexLimits(U8ᵖ(a.Indices), a.IndicesType, 0, uint32(a.IndicesCount))
-						moveClientVBsToVAs(ctx, t, clientVAs, int(limits.Min), int(limits.Max), i, a, s, c, out)
+						moveClientVBsToVAs(ctx, t, clientVAs, limits.First, limits.Count, i, a, s, c, out)
 					}
 
 					glDrawElements := *a
@@ -569,7 +569,7 @@ func compat(ctx context.Context, device *device.Instance) (transform.Transformer
 					data := c.Instances.Buffers[ib].Data.Index(0, s)
 					base := uint32(a.Indices.Address)
 					limits := e.calcIndexLimits(data, a.IndicesType, base, uint32(a.IndicesCount))
-					moveClientVBsToVAs(ctx, t, clientVAs, int(limits.Min), int(limits.Max), i, a, s, c, out)
+					moveClientVBsToVAs(ctx, t, clientVAs, limits.First, limits.Count, i, a, s, c, out)
 				}
 			}
 
@@ -1004,12 +1004,16 @@ func moveClientVBsToVAs(
 	ctx context.Context,
 	t *tweaker,
 	clientVAs map[*VertexAttributeArray]*GlVertexAttribPointer,
-	first, last int, // vertex indices
+	first, count uint32, // vertex indices
 	i atom.ID,
 	a atom.Atom,
 	s *gfxapi.State,
 	c *Context,
 	out transform.Writer) {
+
+	if count == 0 {
+		return
+	}
 
 	rngs := interval.U64RangeList{}
 	// Gather together all the client-buffers in use by the vertex-attribs.
@@ -1030,7 +1034,7 @@ func moveClientVBsToVAs(
 					stride = size
 				}
 				base := memory.Pointer(a.Data) // Always start from the 0'th vertex to simplify logic.
-				rng := base.Range(uint64(last*stride + size))
+				rng := base.Range(uint64(int(first+count-1)*stride + size))
 				interval.Merge(&rngs, rng.Span(), true)
 			}
 		}

--- a/gapis/gfxapi/gles/externs.go
+++ b/gapis/gfxapi/gles/externs.go
@@ -71,7 +71,7 @@ func (e externs) elSize(ty GLenum) uint64 {
 	return uint64(DataTypeSize(ty))
 }
 
-func (e externs) calcIndexLimits(data U8ᵖ, ty GLenum, offset, count uint32) resolve.MinMax {
+func (e externs) calcIndexLimits(data U8ᵖ, ty GLenum, offset, count uint32) resolve.IndexRange {
 	elSize := e.elSize(ty)
 	id := data.Slice(uint64(offset), uint64(offset)+uint64(count)*elSize, e.s).ResourceID(e.ctx, e.s)
 	littleEndian := e.s.MemoryLayout.GetEndian() == device.LittleEndian
@@ -79,7 +79,7 @@ func (e externs) calcIndexLimits(data U8ᵖ, ty GLenum, offset, count uint32) re
 	if err != nil {
 		if errors.Cause(err) == context.Canceled {
 			// TODO: Propagate error
-			return resolve.MinMax{}
+			return resolve.IndexRange{}
 		} else {
 			panic(fmt.Errorf("Could not calculate index limits: %v", err))
 		}
@@ -89,7 +89,7 @@ func (e externs) calcIndexLimits(data U8ᵖ, ty GLenum, offset, count uint32) re
 
 func (e externs) IndexLimits(data U8ᵖ, ty GLenum, offset, count uint32) u32Limits {
 	limits := e.calcIndexLimits(data, ty, offset, count)
-	return u32Limits{First: limits.Min, Last: limits.Max}
+	return u32Limits{First: limits.First, Count: limits.Count}
 }
 
 func (e externs) substr(str string, start, end int32) string {

--- a/gapis/gfxapi/gles/gles.api
+++ b/gapis/gfxapi/gles/gles.api
@@ -675,7 +675,7 @@ sub u32 IndexSize(GLenum indices_type) {
   }
 }
 
-@internal class u32Limits { u32 first u32 last }
+@internal class u32Limits { u32 first u32 count }
 extern u32Limits IndexLimits(u8* indices, GLenum indices_type, u32 offset, u32 count)
 
 /////////////////////////////////////////////////////////////////

--- a/gapis/resolve/index_limits.go
+++ b/gapis/resolve/index_limits.go
@@ -26,16 +26,16 @@ import (
 	"github.com/google/gapid/gapis/service/path"
 )
 
-// MinMax holds a minimum and maximum value.
-type MinMax struct {
-	Min uint32 // Smallest index found in the index buffer.
-	Max uint32 // Largest index found in the index buffer.
+// IndexRange represents the range of indices which were referenced by index buffer.
+type IndexRange struct {
+	First uint32
+	Count uint32
 }
 
-// IndexLimits returns the lowest and highest index contained in the index
+// IndexLimits returns the range of indices which were referenced by index
 // buffer with identifier id. The buffer holds count elements, each of size
 // bytes.
-func IndexLimits(ctx context.Context, data id.ID, count int, size int, littleEndian bool) (*MinMax, error) {
+func IndexLimits(ctx context.Context, data id.ID, count int, size int, littleEndian bool) (*IndexRange, error) {
 	obj, err := database.Build(ctx, &IndexLimitsResolvable{
 		IndexSize:    uint64(size),
 		Count:        uint64(count),
@@ -45,11 +45,14 @@ func IndexLimits(ctx context.Context, data id.ID, count int, size int, littleEnd
 	if err != nil {
 		return nil, err
 	}
-	return obj.(*MinMax), nil
+	return obj.(*IndexRange), nil
 }
 
 // Resolve implements the database.Resolver interface.
 func (c *IndexLimitsResolvable) Resolve(ctx context.Context) (interface{}, error) {
+	if c.Count == 0 {
+		return &IndexRange{First: 0, Count: 0}, nil
+	}
 	min, max := ^uint32(0), uint32(0)
 	data, err := database.Resolve(ctx, c.Data.Id.ID())
 	if err != nil {
@@ -86,5 +89,5 @@ func (c *IndexLimitsResolvable) Resolve(ctx context.Context) (interface{}, error
 		}
 	}
 
-	return &MinMax{Min: min, Max: max}, nil
+	return &IndexRange{First: min, Count: max + 1 - min}, nil
 }

